### PR TITLE
refactor(checkbox): rewrite with hook

### PR DESCRIPTION
**Pull request by hengkx**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     Refactor Checkbox with React hooks.     |
| 🇨🇳 Chinese |   使用 React hooks 重构 Checkbox。        |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/checkbox/Group.tsx [hunk 1]
```diff
@@ -31,9 +31,12 @@ export interface CheckboxGroupProps extends AbstractCheckboxGroupProps {
 }
 
 export interface CheckboxGroupContext {
+  name: string | undefined;
```
Comment (xrkffgg) [inline components/checkbox/Group.tsx:34]: 这个是 必输的？

File: components/transfer/__tests__/__snapshots__/index.test.js.snap [hunk 2]
```diff
@@ -1280,7 +1280,6 @@ exports[`Transfer should support render value and label in item 1`] = `
                     >
                       <Checkbox
                         checked={false}
-                        indeterminate={false}
```
Comment (xrkffgg) [inline components/transfer/__tests__/__snapshots__/index.test.js.snap:1283]: 我看 indeterminate 默认是 false，这个 snap 删除了是预期的吗？

Comment (hengkx) [reply components/transfer/__tests__/__snapshots__/index.test.js.snap:1283]: image
属性加了

**Review by xrkffgg — APPROVED** [review]

**Comment by afc163** [comment]
记得更新一下 https://github.com/ant-design/ant-design/projects/6 ，方面跟踪进度。

**Pull request merged**